File: tests/cxx/isce3/polsar/symmetrize.cpp
#include <complex>

#include <gtest/gtest.h>

#include <isce3/core/Constants.h>
#include <isce3/math/complexOperations.h>
#include <isce3/polsar/symmetrize.h>

using namespace isce3::math::complex_operations;

TEST(PolsarSymmetrizeTest, symmetrize)
{
    using T = float;

    const auto memory_mode_set = {isce3::core::MemoryModeBlockY::SingleBlockY,
            isce3::core::MemoryModeBlockY::MultipleBlocksY};

    const int width = 10, length = 10, nbands = 1;

    isce3::core::Matrix<std::complex<T>> hv_array(length, width);
    isce3::core::Matrix<std::complex<T>> vh_array(length, width);
    isce3::core::Matrix<std::complex<T>> output_array(length, width);

    for (int i = 0; i < length; ++i) {
        for (int j = 0; j < width; ++j) {
            hv_array(i, j) = std::complex<T>(i, j);
            vh_array(i, j) = std::complex<T>(2 * i, 2 * j);
        }
    }

    const int x0 = 0, y0 = 0, band = 1;

    isce3::io::Raster hv_raster("symmetrize_hv_raster.bin", width, length,
            nbands, GDT_CFloat32, "ENVI");

    isce3::io::Raster vh_raster("symmetrize_vh_raster.bin", width, length,
            nbands, GDT_CFloat32, "ENVI");

    isce3::io::Raster output_raster("symmetrize_output_raster.bin", width,
            length, nbands, GDT_CFloat32, "ENVI");

    hv_raster.setBlock(hv_array.data(), x0, y0, width, length, band);
    vh_raster.setBlock(vh_array.data(), x0, y0, width, length, band);

    for (auto memory_mode : memory_mode_set) {

        isce3::polsar::symmetrizeCrossPolChannels(
                hv_raster, vh_raster, output_raster, memory_mode);

        output_raster.getBlock(
                output_array.data(), x0, y0, width, length, band);

        T symmetrization_max_error = 0;

        for (int i = 0; i < length; ++i) {
            for (int j = 0; j < width; ++j) {
                const T error = std::abs(output_array(i, j) * 2 -
                                         (hv_array(i, j) + vh_array(i, j)));
                if (error > symmetrization_max_error) {
                    symmetrization_max_error = error;
                }
            }
        }

        std::cout << "PolSAR symmetrization max. error: "
                  << symmetrization_max_error << std::endl;
        const T symmetrization_error_threshold = 1e-6;
        EXPECT_LT(symmetrization_max_error, symmetrization_error_threshold);
    }
}

int main(int argc, char** argv)
{
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: tests/python/extensions/pybind/polsar/symmetrize.py
#!/usr/bin/env python3
import os
import numpy as np
from osgeo import gdal, gdal_array
import iscetest
import isce3.ext.isce3 as isce3


def _create_raster(outpath, array, width, length, nbands, dtype,
                   driver_name):
    '''
    create and return ISCE3 raster obj. containing given array
    '''
    driver = gdal.GetDriverByName(driver_name)
    dir_name = os.path.dirname(outpath)
    if not os.path.isdir(dir_name):
        os.makedirs(dir_name)
    print('creating test file: ', outpath)
    dset = driver.Create(outpath, width, length, nbands, dtype)
    if dset is None:
        error_message = 'ERROR creating test file: ' + outpath
        raise RuntimeError(error_message)
    dset.GetRasterBand(1).WriteArray(array)
    raster_obj = isce3.io.Raster(outpath)
    return raster_obj

def test_run():
    '''
    run test
    '''

    # set parameters
    width = 10
    length = 10
    nbands = 1
    band = 1
    dtype = gdal.GDT_CFloat32
    memory_mode_list = [
        isce3.core.MemoryModeBlocksY.SingleBlockY,
        isce3.core.MemoryModeBlocksY.MultipleBlocksY] 
    symmetrization_error_threshold = 1e-6
   
    # set file paths
    hv_file = "polsar/symmetrize_hv.bin"
    vh_file = "polsar/symmetrize_vh.bin"
    output_file = "polsar/symmetrize_output.bin"

    # create input arrays
    hv_array = np.zeros((length, width), dtype=np.complex64)
    vh_array = np.zeros((length, width), dtype=np.complex64)
    for i in range(length):
        for j in range(width):
            hv_array[i, j] = i + 1j * j
            vh_array[i, j] = 2 * i + 2j * j 

    # create input raster objects
    hv_raster = _create_raster(hv_file, hv_array, width, length, nbands,
                                    dtype, "ENVI")
    vh_raster = _create_raster(vh_file, vh_array, width, length, nbands,
                                    dtype, "ENVI")

    # iterate over memory modes
    for memory_mode in memory_mode_list:

        # create output raster object
        output_raster = isce3.io.Raster(output_file, width, length, nbands,
                                    dtype, "ENVI")

        # symmetrize cross-pol channels
        isce3.polsar.symmetrize_cross_pol_channels(
            hv_raster, vh_raster, output_raster, memory_mode)
        del output_raster

        # read output raster
        ds = gdal.Open(output_file, gdal.GA_ReadOnly)
        output_array = ds.GetRasterBand(band).ReadAsArray()

        # compute max error
        symmetrization_max_error = np.max(np.abs(output_array[i, j] - 
                                    (hv_array[i, j] + vh_array[i, j]) / 2.0))

        # evaluate max error
        print('PolSAR symmetrization max. error:', symmetrization_max_error)
        assert(symmetrization_max_error  < symmetrization_error_threshold )


if __name__ == "__main__":
    test_run()
